### PR TITLE
Fix binary operations with 1st source operand a constant

### PR DESCRIPTION
**Pull request by usha1830**
DPDK needs the binary operation to have 1st src operand same as dst operand.
Some of the existing cases were not handled by the existing transformation pass.
`a = 10-a`
should be converted to 
```
tmp = 10
tmp = tmp - a
a = tmp
```

### CODE REVIEW

File: backends/dpdk/dpdkArch.cpp [hunk 1]
```diff
@@ -1086,10 +1096,30 @@ ConvertBinaryOperationTo2Params::postorder(IR::AssignmentStatement *a) {
...
-            bin_expr->right = r->left;
-            a->right = bin_expr;
-            return a;
+            if (isCommutativeBinaryOperation(r)) {
+                bin_expr->left = r->right;
+                bin_expr->right = r->left;
+                a->right = bin_expr;
+                return a;
+            } else {
+                // Expressions like "a = 10 - a" should be replaced with the following block
```
Comment (kamleshbhalui) [inline backends/dpdk/dpdkArch.cpp:1068]: will it work if we convert to `a = -a + 10`

Comment (usha1830) [reply backends/dpdk/dpdkArch.cpp:1068]: I don't think so. Negation is another unary operation and would need an assignment like tmp =-a (Dpdk does not support any complex operations)
Also, handling such a way will not handle the other non-commutative operations like shl, shr. We will then have to find inverse of all such operations which will unnecessarily complicate the implementation here.

Comment (kamleshbhalui) [reply backends/dpdk/dpdkArch.cpp:1068]: is p4 ir not in ssa form?

File: backends/dpdk/dpdkArch.cpp [hunk 2]
```diff
@@ -1086,10 +1096,30 @@ ConvertBinaryOperationTo2Params::postorder(IR::AssignmentStatement *a) {
...
+                // a = tmp;
+                IR::IndexedVector<IR::StatOrDecl> code_block;
+                auto control = findOrigCtxt<IR::P4Control>();
+                auto parser = findOrigCtxt<IR::P4Parser>();
+                auto tmpOp1 = new IR::PathExpression(IR::ID(refMap->newName("tmp")));
+                injector.collect(control, parser,
+                                 new IR::Declaration_Variable(tmpOp1->path->name, left->type));
+                code_block.push_back(new IR::AssignmentStatement(tmpOp1, r->left));
+                bin_expr->left = tmpOp1;
+                bin_expr->right = r->right;
```
Comment (kamleshbhalui) [inline backends/dpdk/dpdkArch.cpp]: is it redundant since it was cloned?

Comment (usha1830) [reply backends/dpdk/dpdkArch.cpp]: Yes. I will update this.

File: backends/dpdk/dpdkArch.cpp (hunk 1 above)
Comment (usha1830) [reply backends/dpdk/dpdkArch.cpp:1068]: No. I don't think so.

File: backends/dpdk/dpdkArch.cpp (hunk 2 above)
Comment (usha1830) [reply backends/dpdk/dpdkArch.cpp]: Done.

**Comment by usha1830** [comment]
I will take care of cleaning the commit history while creating public PR

**Review by kamleshbhalui — APPROVED** [review]

**Review by github-sajan — APPROVED** [review]
Looks fine to me

**Pull request merged**